### ISSUE TITLE
cAdvisor should retrieve application metrics using container's network namespace

### ISSUE THREAD

**Issue by DirectXMan12** [google/cadvisor#1212]
cAdvisor retrieves application metrics by issuing a GET request to an HTTP endpoint configured by placing a configuration file in the container's filesystem.  The endpoint generally looks something like "http://localhost:8080/metrics".

However, when cAdvisor issues the GET request, it does so in it's own network namespace (in the case of Kubelet's cAdvisor, this is generally the host's network namespace).  Ergo, when exposing custom metrics, each container must have its own unique host port to serve the metrics on.

This problematic, since, in Kubernetes, it means that the user launching the post must a) expose the metrics on a hostPort, and b) ensure that hostPort does conflict with any other pod's metrics port or any ports used by processes on the host.

Instead, cAdvisor should perform the HTTP request in the container's network namespace (e.g. by calling out to nsenter).

**Comment by DirectXMan12** [comment]
(I'm willing to work on this -- the fix shouldn't be too hard)

**Comment by DirectXMan12** [comment]
cc @mwielgus @piosz

**Comment by vishh** [comment]
Switching network namespaces has error prone in the past. Is it fair to
expect the metrics to be available in the container/pod IP:port?

**Comment by piosz** [comment]
cc @jszczepkowski

**Comment by DirectXMan12** [comment]
@vishh you *could* put in on the pod IP, but then you'd have to know the pod IP when writing the configuration, so you'd basically be back to have kube write the configuration.  This is suboptimal for that reason, but also because it means that you'd have to expose the metrics externally.  If you just expose them on localhost in the pod's network namespace, it means that others can't easily get at the metrics (which is desirable in many cases).

**Comment by vishh** [comment]
> @vishh you could put in on the pod IP, but then you'd have to know the pod IP when writing the configuration, so you'd basically be back to have kube write the configuration. 

Depends on when the configuration is added. It is binded at the last minute, the kubelet can generate configs dynamically for example. We could also compose the configuration and have provisions for overriding specific fields in the config. 

In addition to that, we can have cadvisor identify the IP of a given container in a pod as well.

**Comment by vishh** [comment]
@DirectXMan12 To be clear, we have tried switching namespaces and it has caused bugs in the past. Otherwise your idea of exposing stats only on localhost does make a lot of sense.

**Comment by DirectXMan12** [comment]
> Depends on when the configuration is added. It is binded at the last minute, the kubelet can generate configs dynamically for example.

With the exception of the "exposing metrics to anyone with the pod IP" issue mention above, this seems pretty reasonable to me. However, I will note that this was originally part of how custom metrics worked, and I recall that there were objections to having Kubelet autogenerate custom metrics config on the fly.

For example, you could have annotations like `custom-metrics/port=8080`, `custom-metrics/path=/metrics`, `custom-metrics/metrics=foo,bar` (or something similar as one object), with the path defaulting to `/metrics` and `/metrics` defaulting to "*" if those weren't specified.

> We could also compose the configuration and have provisions for overriding specific fields in the config.

I prefer the "generate on the fly" idea from above.  It would probably be easier/fewer moving parts for users to work with anyway.

> To be clear, we have tried switching namespaces and it has caused bugs in the past. Otherwise your idea of exposing stats only on localhost does make a lot of sense.

Ah, ok.  I'd be curious to hear what they were -- I was thinking something very simply like shelling out to `nsenter --target $PID --net -- curl $URL`.

**Comment by vishh** [comment]
> With the exception of the "exposing metrics to anyone with the pod IP" issue mention above, this seems pretty reasonable to me. However, I will note that this was originally part of how custom metrics worked, and I recall that there were objections to having Kubelet autogenerate custom metrics config on the fl

There are multiple means to achieve this. For example, a new feature called "init containers" can be used to dynamically generate configuration for cAdvisor. We could support variable substitution and get pod IP from env vars.

> Ah, ok. I'd be curious to hear what they were

Go threads were in random network namespaces. On top of that exec'ing is also not cheap and adds latency.

**Comment by DirectXMan12** [comment]
> Go threads were in random network namespaces.

Yeah, that's mainly why I wasn't thinking of just using the system calls, etc.

> On top of that exec'ing is also not cheap and adds latency.

Fair.  It seems like the "expose metrics on the pod IP" is the way we should go for the moment, then.

> For example, a new feature called "init containers" can be used to dynamically generate configuration for cAdvisor.  We could support variable substitution and get pod IP from env vars.

So, I think I'd prefer just having the kubelet generate the config.  Somewhat separate from this issue, I've a suspicion that the whole configmap-volume-manually-writing-json workflow will end up being confusing and complicated for user.  If we switched to just having Kubelet generate the config, we could kill two birds with one stone here, so to speak.

My worry with the "init containers" idea is that a) it appears to just be at the proposal phase, and b) unless the init container is generated automatically (at which point why not just have kubelet generate the config directly), it makes things more complicated, not less (see above).  Thus, if we can't just have kubelet generate the config, I'd probably lean towards env vars.

**Comment by vishh** [comment]
> So, I think I'd prefer just having the kubelet generate the config.
Somewhat separate from this issue, I've a suspicion that the whole
configmap-volume-manually-writing-json workflow will end up being confusing
and complicated for user. If we switched to just having Kubelet generate
the config, we could kill two birds with one stone here, so to speak.

We can create a generic cadvisor plugin container that generates configs
for most open source applications.

> My worry with the "init containers" idea is that a) it appears to just be
at the proposal phase, and b) unless the init container is generated
automatically (at which point why not just have kubelet generate the config
directly), it makes things more complicated, not less (see above). Thus, if
we can't just have kubelet generate the config, I'd probably lean towards
env vars.

Reg `a` - @smarterclayton is working on it actively.
Reg `b` - Generating configs is a problem that is common to many apps. We
should try to build and support common patterns. We can create a standard
cAdvisor configuration container that users can run. I can also imagine
cluster admins injecting init containers into pods as part of admission
control.

**Comment by xiangpengzhao** [comment]
@vishh Does cAdvisor support pod IP:port now? If so, how?
Another question, what if the node can't access the pod IP?
Thanks!

**Comment by DirectXMan12** [comment]
IIRC, yes, cAdvisor now supports pod IP:port now.

**Comment by xiangpengzhao** [comment]
@DirectXMan12 Thanks for your information! 
Any idea about the question `what if the node can't access the pod IP`? In Kubernetes environments, the node network and the pod network may be isolated (depends on the third-party CNI network plugin). Then kubelet (cAdvisor) can't access the pod IP. In this case, how to collect custom metrics since we can neither access the pod IP nor use container's network namespace ? Thanks!

**Comment by alex88** [comment]
@DirectXMan12 @xiangpengzhao it supports pod IP:port, how? with the init containers?

**Comment by piosz** [comment]
This is out of the scope of Heapster. It doesn't collect application level metrics.

**Comment by xiangpengzhao** [comment]
@alex88  Seems like config like this file [sample_config_endpoint_config](https://github.com/google/cadvisor/blob/master/collector/config/sample_config_endpoint_config.json), then cAdvisor will access pod IP:port. Not so sure about this.

**Comment by cirocosta** [comment]
Hey, sorry if I'm getting a little off the main track of the issue but I have a question: 

- Is this configuration restricted to docker bridge networks and kubernetes pods? 

As I see  there's no way to get a response from `<ContainerIp>:<Port>` without a bridge / port binding / kubernetes pod networking without entering the namespace (or am I wrong? 🤔 ).

From the PR that adds support there's the addition of the following:

https://github.com/google/cadvisor/blob/6ef612f21e042dcbec189ab253741f94e47804d7/collector/util.go#L21-L25

but this is just gathering of the `ip` address and no collection of `netns`. 

Please let me know if I'm missing something!

Thx!

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.
